runtime/cudaq/qis/managers/photonics/PhotonicsExecutionManager.cpp

```diff
@@ -0,0 +1,258 @@
+        if (term1 == 0) {
+          continue;
+        }
+        double term2 = pow((-1), (l)) *
+                       (sqrt(_fast_factorial(n1)) * sqrt(_fast_factorial(n2)) *
+                        sqrt(_fast_factorial(N1)) * sqrt(_fast_factorial(N2)));
+        double term3 = (_fast_factorial(k) * _fast_factorial(n1 - k) *
+                        _fast_factorial(l) * _fast_factorial(n2 - l));
+        double term = term1 * term2 / term3;
+        sum += term;
+      } else {
+        continue;
+      }
+    }
+
+    return sum;
+  }
+
+  void beamsplitter(const double theta, qpp::cmat &BS) {
+    int d = sqrt(BS.rows());
+    //     """Returns a matrix representing a beam splitter
+    for (int n1 = 0; n1 < d; ++n1) {
+      for (int n2 = 0; n2 < d; ++n2) {
+        int nxx = n1 + n2;
+        int nxd = std::min(nxx + 1, d);
+        for (int N1 = 0; N1 < nxd; ++N1) {
+          int N2 = nxx - N1;
+          if (N2 >= nxd) {
+            continue;
+          } else {
+
+            BS(n1 * d + n2, N1 * d + N2) =
+                _calc_beamsplitter_elem(N1, N2, n1, n2, theta);
+          }
+        }
+      }
+    }
+  }
+
+public:
+  PhotonicsExecutionManager() {
+
+    instructions.emplace("plusGate", [&](const Instruction &inst) {
+      auto &[gateName, params, controls, qudits] = inst;
+      auto target = qudits[0];
+      int d = target.levels;
+      qpp::cmat u{qpp::cmat::Zero(d, d)};
+      u(0, d - 1) = 1;
+      for (int i = 1; i < d; i++) {
+        u(i, i - 1) = 1;
+      }
+      cudaq::info("Applying plusGate on {}<{}>", target.id, target.levels);
+      state = qpp::applyCTRL(state, u, {}, {target.id}, target.levels);
+    });
+
+    instructions.emplace("beamSplitterGate", [&](const Instruction &inst) {
+      auto &[gateName, params, controls, qudits] = inst;
+      auto target1 = qudits[0];
+      auto target2 = qudits[1];
+      size_t d = target1.levels;
+      const double theta = params[0];
+      qpp::cmat BS{qpp::cmat::Zero(d * d, d * d)};
+      beamsplitter(theta, BS);
+      cudaq::info("Applying beamSplitterGate on {}<{} and {}<{}>", target1.id,
+                  target1.levels, target2.id, target2.levels);
+      state = qpp::apply(state, BS, {target1.id, target2.id}, d);
+    });
+
+    instructions.emplace("phaseShiftGate", [&](const Instruction &inst) {
+      auto &[gateName, params, controls, qudits] = inst;
+      auto target = qudits[0];
+      size_t d = target.levels;
+      const double phi = params[0];
+      qpp::cmat PS{qpp::cmat::Identity(d, d)};
+      const std::complex<double> i(0.0, 1.0);
+      for (size_t n = 0; n < d; n++) {
+        PS(n, n) = std::exp(n * phi * i);
+      }
+      cudaq::info("Applying phaseShiftGate on {}<{}>", target.id,
+                  target.levels);
+      state = qpp::applyCTRL(state, PS, {}, {target.id}, target.levels);
+    });
+  }
+
+  virtual ~PhotonicsExecutionManager() = default;
+
+  // cudaq::SpinMeasureResult measure(cudaq::spin_op &op) override {
```

Comment:
dead code. You may want to turn this on and throw an exception indicating that spin_op observation (cudaq::observe()) is not supported for this simulator.

Reply:
Throwing exception